### PR TITLE
session: add a `doNotCommit` flag to transaction when `StmtCommit` fail

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve? 

If `StmtCommit` meets error, it should not leave dirty data in Txn.
When the operation is irreversible, it's better to mark the whole Txn as failed.

### What is changed and how it works?

Add a `doNotCommit` flag.
If a Txn has this flag, it acts as a normal Txn, except that it can't Commit. 

### Check List 

Tests 

 - Unit test

Side effects

 - Increased code complexity

@winkyao

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/8918)

### CODE REVIEW

File: session/session_fail_test.go [hunk 1]
```diff
@@ -22,17 +22,24 @@ import (
...
-	tk.MustQuery(`select * from t`).Check(testkit.Rows("1"))
+
+	gofail.Disable("github.com/pingcap/tidb/session/mockStmtCommitError")
+
+	tk.MustExec("insert into t values (3)")
+	tk.MustExec("insert into t values (4)")
+	_, err = tk.Exec("commit")
+	c.Assert(err, NotNil)
+
+	tk.MustQuery(`select * from t`).Check(testkit.Rows())
```
Comment (winkyao) [inline session/session_fail_test.go]: Please add a test that error occurred during the partial StmtCommit to Transaction, and make sure there is no data consistent problem.

Comment (tiancaiamao) [reply session/session_fail_test.go]: Now, line 32 mock partial of the data fail...

**Pull request merged**